File: orion-viewer/src/universe/constellation/orion/viewer/prefs/OrionBookPreferences.java
/*
 * Orion Viewer - pdf, djvu, xps and cbz file viewer for android devices
 *
 * Copyright (C) 2011-2013  Michael Bogdanov & Co
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package universe.constellation.orion.viewer.prefs;

import android.os.Bundle;
import android.preference.ListPreference;
import android.preference.PreferenceActivity;
import android.preference.PreferenceCategory;
import universe.constellation.orion.viewer.R;

/**
 * User: mike
 * Date: 17.05.12
 * Time: 21:41
 */
public class OrionBookPreferences extends PreferenceActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        getOrionContext().applyTheme(this);

        super.onCreate(savedInstanceState);

        addPreferencesFromResource(R.xml.book_preference);

        PreferenceCategory general = (PreferenceCategory) getPreferenceScreen().findPreference("GENERAL");
        ListPreference screenOrientation = (ListPreference) general.findPreference("screenOrientation");

        boolean isLevel9 = getOrionContext().getSdkVersion() >= 9;


        CharSequence[] values = getResources().getTextArray(isLevel9 ? R.array.screen_orientation_full_desc : R.array.screen_orientation_desc);
        CharSequence[] newValues = new CharSequence[values.length];
        for (int i = 0; i < values.length; i++) {
            newValues[i] = values[i];
        }
        newValues[0] = getResources().getString(R.string.orientation_default_rotation);
        screenOrientation.setEntries(newValues);

        if (isLevel9) {
            screenOrientation.setEntryValues(R.array.screen_orientation_full);
        }
//            if (!isLevel9) {
//                CharSequence [] entries = screenOrientation.getEntries();
//                CharSequence [] values = screenOrientation.getEntryValues();
//                entries = Arrays.copyOf(entries, 3);
//                values = Arrays.copyOf(values, 3);
//                screenOrientation.setEntries(entries);
//                screenOrientation.setEntryValues(values);
//            }

    }


    public OrionApplication getOrionContext() {
        return (OrionApplication) getApplicationContext();
    }

}

File: orion-viewer/src/main/java/universe/constellation/orion/viewer/prefs/OrionBookPreferences.kt
/*
 * Orion Viewer - pdf, djvu, xps and cbz file viewer for android devices
 *
 * Copyright (C) 2011-2017 Michael Bogdanov & Co
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package universe.constellation.orion.viewer.prefs

import android.os.Bundle
import android.preference.PreferenceScreen
import universe.constellation.orion.viewer.R
import universe.constellation.orion.viewer.R.array.*
import universe.constellation.orion.viewer.R.string.*
import universe.constellation.orion.viewer.android.DSLPreferenceActivity
import universe.constellation.orion.viewer.prefs.BookPreferenceKey.*


class OrionBookPreferences : DSLPreferenceActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        (applicationContext as OrionApplication).applyTheme(this)

        super.onCreate(savedInstanceState)

        rootScreen {
            bookPreferences(this, false)
        }
    }


    companion object {
        fun DSLPreferenceActivity.bookPreferences(preferenceScreen: PreferenceScreen, isGeneral: Boolean) {
            preferenceScreen.category {
                title = (if (!isGeneral) book_pref_title else pref_default_book_setting).stringRes

                preference<OrionListPreference> {
                    isCurrentBookOption = !isGeneral
                    orionKey = SCREEN_ORIENTATION
                    title = pref_screen_orientation.stringRes
                    summary = pref_book_screen_orientation.stringRes
                    dialogTitle = pref_screen_orientation.stringRes
                    setDefaultValue("DEFAULT")

                    screen_orientation_full_desc.stringArray.let {
                        if (!isGeneral) {
                            it[0] = orientation_default_rotation.stringRes
                        }
                        entries = it
                    }
                    entryValues = screen_orientation_full.stringArray
                }

                if (isGeneral) {
                    preference<OrionListPreference> {
                        isCurrentBookOption = !isGeneral
                        orionKey = ZOOM
                        title = pref_bookDefaultZoom.stringRes
                        summary = pref_bookDefaultZoom.stringRes
                        dialogTitle = pref_bookDefaultZoom.stringRes
                        setDefaultValue("0")

                        entries = default_zoom_option_desc.stringArray
                        entryValues = default_zoom_option.stringArray
                    }
                }

                preference<OrionLayoutDialog> {
                    isCurrentBookOption = !isGeneral
                    orionKey = PAGE_LAYOUT
                    title = pref_page_layout.stringRes
                    summary = pref_page_layout.stringRes
                    dialogTitle = pref_page_layout.stringRes
                    setDefaultValue(0)
                }

                preference<OrionListPreference> {
                    isCurrentBookOption = !isGeneral
                    orionKey = WALK_ORDER
                    title = pref_walk_order.stringRes
                    summary = pref_walk_order.stringRes
                    dialogTitle = pref_walk_order.stringRes
                    setDefaultValue(ABCD.stringRes)
                    setDialogIcon(R.drawable.walk_order)

                    entries = walk_orders_desc.stringArray
                    entryValues = walk_orders.stringArray
                }

                preference<OrionListPreference> {
                    isCurrentBookOption = !isGeneral
                    orionKey = COLOR_MODE
                    title = pref_color_mode.stringRes
                    summary = pref_color_mode.stringRes
                    dialogTitle = pref_color_mode.stringRes
                    setDefaultValue("CM_NORMAL")

                    entries = color_mode_desc.stringArray
                    entryValues = color_mode.stringArray
                }

                preference<OrionListPreference> {
                    isCurrentBookOption = !isGeneral
                    orionKey = CONTRAST
                    title = pref_book_contrast.stringRes
                    summary = pref_book_contrast_desc.stringRes

                    val values = Array(17) { index ->
                        when {
                            index <= 5 -> 10 + index * 15
                            index <= 12 -> 100 + (index - 6) * 50
                            else -> 500 + (index - 12) * 100
                        }.toString()
                    }

                    entries = values
                    entryValues = values
                    setDefaultValue("100")
                }

                if (!isGeneral) {
                    preference<SeekBarPreference> {
                        isCurrentBookOption = !isGeneral
                        orionKey = THRESHOLD
                        title = pref_book_threshold.stringRes
                        summary = pref_book_threshold_desc.stringRes

                        minValue = 1
                        maxValue = 255
                        defaultSeekValue = 255
                    }
                }
            }
        }
    }
}


